### ISSUE TITLE
 TypeError: Object [object SequelizeInstance] has no method 'getUsers'

### ISSUE THREAD

**Issue by chuks** [sequelize/sequelize#3362]
Brilliant library, just saw it this weekend, and already paying around with it. But...I've got some major roadblocks. In my schema, I have a many-many association in my models, and cannot call the instance method getUsers. I'm using Sequelize 2.0.4 . Here are my models:

--User.js
"use strict";

var definition = require(__dirname + '/./../utils/models/definitions/UserDefs.js').definition;  (ps: this loads the fields as an object. I just decided to put it in another Js file for brevity)

module.exports = function(sequelize, DataTypes) {
    console.log("user model - definition is: "+definition);

    var User = sequelize.define('User', definition, {
        classMethods: {
            associate: function(models) {

                User.belongsToMany(models.Happening, {
                    through: 'UserHappening',
                    foreignKey: 'User_id'
                });
            }
        },
        freezeTableName: true,
        tableName: 'users'
    });

  return User;
};

--Happenings.js
"use strict";
var definition = require(__dirname + '/./../utils/models/definitions/HappeningDefs.js').definition;

module.exports = function(sequelize, DataTypes) {
    console.log("Happening model definition is: "+definition);

    var Happening = sequelize.define('Happening', definition, {
        classMethods: {
            getDefinition: function() {
                return definition;
            },
            associate: function(models) {

                Happening.belongsToMany(models.User, {
                    through: 'UserHappening',
                    foreignKey: 'Happening_id'
                });
            }
        },
        freezeTableName: true,
        tableName: 'happennings'
    });

 	return Happening;
};

--UserHappenings.js  (a custom through model that ties User and Happening models together)
"use strict";
module.exports = function(sequelize, DataTypes) {
	var UserHappennings = sequelize.define('userhappennings',{
	 	hit: {
	 		type: DataTypes.BOOLEAN,
	 		defaultValue: false
	 	},
	    view: {
	 		type: DataTypes.BOOLEAN,
	 		defaultValue: false
	 	},
	 	calendar: {
	 		type: DataTypes.BOOLEAN,
	 		defaultValue: false
	 	},
	});	 	

	return UserHappennings;
};	

Now, in my controller, i try to do this:

exports.getAllHappenings = function(req, res) {
    console.log("happeningController: getAllHappenings");

    models.Happening.findAll({})
    .then(function(happenings){
        console.log("happeningController: getHappeningById - successfully retrieved happenings: "+happenings);
    
         //ERROR comes from here. If i try to do something like this:
         happenings.forEach(function(happening){  <= I have also tried calling getUsers() on happenings
                happening.getUsers().then(function (users) {        <= Why can't I call getUsers()?
                  });
         });
......rest of code truncated/irrelevant

//---------
Here is really what I want to accomplish
1 - The Users belonging to a Happening can fall into one or more of "hits", "views" and "calendar" categories, as described in UserHappenings though model.
2 - How do I add a User to Happening, under one or more of the the above categories defined in UserHappenings?
3 - how can I get access to the Users associated with the happening object, that satisfy the above condition?
4 - I would like to include this array of Users in the return happenings array
5 - I would like to further separate the Users array into the three categories as defined in the UeerHappenings through model, i.e. Users where "hits" is true, Users where "views" is true, and Users where "calendar" is true.

**Comment by mickhansen** [comment]
Are you sure that the associate method is being called?
Can you try logging `Happening.Instance.prototype`?

**Comment by chuks** [comment]
Thanks fro the response. How can i verify that it is being called? Since it's a class method, doesn't it get called by default when it hits create, or do I have to explicitly call associate? The tables are created in the database alright. I'fe dropped the tabled several times, and ran it, and each time, the tables corresponding to the models are generated. My setup is exactly as described here - https://github.com/sequelize/express-example

**Comment by mickhansen** [comment]
It's a static method, just means it's exposed on the class as a property (i.e. Model.associate), you still have to call it :)

You can place a console.log inside the method to ensure it's being called.

**Comment by chuks** [comment]
Ok, got it. Thanks. How and when do I call it? I am using the boilerplate code in the link I posted above. I believe the point at which this is called is index.js locate inside models directory. And then in bin/www, I am calling models.sequelize.sync()...

Here is the listing of index.js:

var db = {};
fs
  .readdirSync(__dirname)
  .filter(function(file) {
    return (file.indexOf(".") !== 0) && (file !== "index.js");
  })
  .forEach(function(file) {
    var model = sequelize["import"](path.join(__dirname, file));
    db[model.name] = model;
  });

Object.keys(db).forEach(function(modelName) {
  if ("associate" in db[modelName]) {
    db[modelName].associate(db);
  }
});

db.sequelize = sequelize;
db.Sequelize = Sequelize;

module.exports = db;

**Comment by mickhansen** [comment]
That code should be calling the method, you can verify by placing a logging method inside.
Did you ever log `Happening.Instance.prototype`?

**Comment by chuks** [comment]
I put in a console.log in the associate method, and it did not spit out anything. Apparently, it is not getting called. Where can i put in that Happening.Instance.prototype log?

**Comment by chuks** [comment]
At what point does this method gets called during the initialization process, using the boiler plate code above? Apparently, it's not being called, and i have used the code as is.

**Comment by mickhansen** [comment]
The boilerplate should be looking through your models, checking if they have an associate property, and if they do call it.

**Comment by mickhansen** [comment]
https://github.com/sequelize/express-example/blob/master/models/index.js#L22 Try throwing some logging statements around, see what happens.

**Comment by chuks** [comment]
I there in some logs inside that for loop, and for these:

associate. Happening.Instance.prototype: [object SequelizeInstance]
associate. User.Instance.prototype: [object SequelizeInstance]
associate. Happening.Instance.prototype: [object SequelizeInstance]
associate. User.Instance.prototype: [object SequelizeInstance]
associate. Happening.Instance.prototype: [object SequelizeInstance]
associate. User.Instance.prototype: [object SequelizeInstance]
associate. Happening.Instance.prototype: [object SequelizeInstance]
associate. User.Instance.prototype: [object SequelizeInstance]
associate. Happening.Instance.prototype: [object SequelizeInstance]
associate. User.Instance.prototype: [object SequelizeInstance]

So it looks like it found those. but the logs inside the actual static methods of the models are not getting called.

**Comment by mickhansen** [comment]
No idea what's going on here, can you try logging `db[modelName].associate`?

**Comment by chuks** [comment]
Here is what I get: 

associate: db[modelName].associate: function (models) {

                Happening.belongsToMany(models.User, {
                    foreignKey: 'user_id',
                    through: 'UserHappennings',
                    as: 'Hits'
                });

                Happening.belongsToMany(models.User, {
                    foreignKey: 'user_id',
                    through: 'UserHappennings',
                    as: 'Views'
                });

                Happening.belongsToMany(models.User, {
                    foreignKey: 'user_id',
                    through: 'UserHappennings',
                    as: 'AddedToCalendar'
                });

                /*Happening.hasMany(models.User, {
                    through: 'UserHappennings',
                    foreignKey: 'user_id',
                    onDelete: 'SET NULL', 
                    onUpdate: 'CASCADE',
                    as: 'Views'
                });

                Happening.hasMany(models.User, {
                    through: 'UserHappennings',
                    foreignKey: 'user_id',
                    onDelete: 'SET NULL', 
                    onUpdate: 'CASCADE',
                    as: 'AddedToCalendar'
                });*/
            }
associate. Happening.Instance.prototype: [object SequelizeInstance]
associate. User.Instance.prototype: [object SequelizeInstance]
associate: db[modelName].associate: function (models) {

                HappeningView.belongsTo(models.User, {
                    foreignKey: 'happening_view_id'
                });

            }
associate. Happening.Instance.prototype: [object SequelizeInstance]
associate. User.Instance.prototype: [object SequelizeInstance]
associate: db[modelName].associate: function (models) {

                Happeningtest.belongsToMany(models.Usertest, {
                    through: 'UserHappening',
                    foreignKey: 'Happeningtest_id'
                });
            }
associate. Happening.Instance.prototype: [object SequelizeInstance]
associate. User.Instance.prototype: [object SequelizeInstance]
associate: db[modelName].associate: function (models) {

                User.belongsToMany(models.Happening, {
                    through: 'UserHappening',
                    foreignKey: 'User_id'
                });
            }
associate. Happening.Instance.prototype: [object SequelizeInstance]
associate. User.Instance.prototype: [object SequelizeInstance]
associate: db[modelName].associate: function (models) {

                Usertest.belongsToMany(models.Happeningtest, {
                    through: 'UserHappening',
                    foreignKey: 'User_id'
                });
            }

**Comment by mickhansen** [comment]
Are you sure you put logging statements inside associate? It doesn't look like it from that log.

**Comment by chuks** [comment]
I believe I am. Here is Happening.log

"use strict";
var definition = require(__dirname + '/./../utils/models/definitions/HappeningDefs.js').definition;

module.exports = function(sequelize, DataTypes) {
    console.log("Happening model definition is: "+definition);

    var Happening = sequelize.define('Happening', definition, {
        classMethods: {
            getDefinition: function() {
                return definition;
            },
            associate: function(models) {
                console.log("Happening - associate. Happening.Instance.prototype: "+Happening.Instance.prototype);
                Happening.belongsToMany(models.User, {
                    through: 'UserHappening',
                    foreignKey: 'Happening_id'
                });
            }
        },
        freezeTableName: true,
        tableName: 'happennings'
    });

 	return Happening;
};

**Comment by chuks** [comment]
Is there another way I can explicitly call this method? I can't figure out why it is not being called

**Comment by mickhansen** [comment]
You can try  `Happening.associate()`

**Comment by chuks** [comment]
So, here is what I did to solve the problem. I had other models in my models folder. Now, i was't using those yet. I deleted those, and then all of a sudden, associate from happenings started being hit by that for loop. Beats me! It's ok for now, but I wonder if I'll run into the same problem when I introduce other models. Anyway, that was a quick fix, so there's no telling if the other models i deleted were wrong or what not. Thanks for your help...got me to this point, and it's working now. 

On a different note, it's turned out that I may not even need the associations! Ha! Anyway, the good news is that I have a better understanding of the library, for future reference. I think keeping my schema simple enough will get rid of the need to do complex associations. Thanks again!
